### PR TITLE
Implement linear referencing functions with tests

### PR DESCRIPTION
**Pull request by sharkAndshark**

### CODE REVIEW

File: LinqToDBPostGisNetTopologySuite/LinearReferencing.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        /// <remarks>
+        /// See https://postgis.net/docs/manual-3.0/ST_LocateBetween.html
+        /// </remarks>
+        /// <param name="geometry">Input geometry</param>
+        /// <param name="measure_start">Measure start</param>
+        /// <param name="measure_end">Measure end</param>
+        /// <param name="offset">Offset</param>
+        /// <returns>Derived geometry collection with elements that match the specified range of measures inclusively</returns>
+        [Sql.Function("ST_LocateBetween", ServerSideOnly = true)]
+        public static NTS.GeometryCollection STLocateBetween(this NTSG geometry, double measure_start, double measure_end, double offset)
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite/LinearReferencing.cs]: - Returning common type 'NetTopologySuite.Geometries.Geometry' (`NTSG`) would be OK (as all methods are implemented now: PostGIS `geometry` maps to `NetTopologySuite.Geometries.Geometry`)
- `measure_start`, `measure_end` are violating naming convention for C# (should be `measureStart`, `measureEnd`)

File: LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+            using (var db = new PostGisTestDataConnection(TestDatabaseConnectionString))
+            {
+                db.TestGeometries.Delete();
+            }
+        }
+
+        [Test]
+        public void TestSTLineInterpolatePoint()
+        {
+            var geomText = "LINESTRING(25 50, 100 125, 150 190)";
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs]: Would be better
`const string ewkt = ...` or `const string EWKT = ...`

File: LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+                Assert.AreEqual(3.0, result.Coordinates[0].Z, 1.0E-8);
+
+                Assert.AreEqual(2.0, result.Coordinates[1].X, 1.0E-8);
+                Assert.AreEqual(3.0, result.Coordinates[1].Y, 1.0E-8);
+                Assert.AreEqual(4.0, result.Coordinates[1].Z, 1.0E-8);
+            }
+        }
+
+        [Test]
+        public void TESTSTInterpolatePoint()
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs]: Naming convention for functions is PascalCase:
`TestSTInterpolatePoint`

File: LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+                .Value(g => g.Geometry, () => GeometryInput.STGeomFromEWKT(geomLineText))
+                .Insert();
+
+                var result = db.TestGeometries
+                .Where(g => g.Id == 1)
+                .Select(g => g.Geometry.STInterpolatePoint(GeometryInput.STGeomFromEWKT(geomPointText)))
+                .Single() as double?;
+
+                Assert.IsNotNull(result);
+                Assert.AreEqual(10.0, result.Value, 1.0E-8);
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs:289]: Simple checks for `null` arguments for better test coverage, like this:
````c#
Assert.IsNull(db.Select(() => LinearReferencing.STInterpolatePoint(null, null)));
````

File: LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+                .Value(g => g.Id, 1)
+                .Value(g => g.Geometry, () => GeometryInput.STGeomFromEWKT(geomText))
+                .Insert();
+
+                var result = db.TestGeometries
+                .Where(g => g.Id == 1)
+                .Select(g => g.Geometry.STLineInterpolatePoint(0.20))
+                .Single() as NTSG.Point;
+
+                Assert.AreEqual(51.5974135047432, result.X, 1.0E-8);
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs:37]: Be careful to copy these values from PostGIS Reference Examples; it sometimes differs from own results (may be due another PostgreSql / PostGIS versions). I'll think, how to use more obvious values in tests.

File: LinqToDBPostGisNetTopologySuite/LinearReferencing.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        }
+
+        /// <summary>
+        /// Return the value of the measure dimension of a geometry at the point closed to the provided point
+        /// </summary>
+        /// <remarks>
+        /// See https://postgis.net/docs/manual-3.0/ST_InterpolatePoint.html
+        /// </remarks>
+        /// <param name="geometry">Input geometry</param>
+        /// <param name="point">Elevation_start</param>
```
Comment (apdevelop) [inline LinqToDBPostGisNetTopologySuite/LinearReferencing.cs]:
````c#
<param name="point">Given point</param>
````

File: LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs (hunk 5 above)
Comment (sharkAndshark) [reply LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs:37]: Some simple test cases with short expects (integer woubld be better )may be helpful?

File: LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs (hunk 3 above)
Comment (sharkAndshark) [reply LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs]: Will search some vs/code extensions' help.   😂 Actually the code has been through many times review by myself.....

File: LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs (hunk 5 above)
Comment (apdevelop) [reply LinqToDBPostGisNetTopologySuite.Tests/LinearReferencingTests.cs:37]: Not sure about integers, but test case with `STLineInterpolatePoint(0.5)` and asserting like this
````c#
Assert.AreEqual((100.0-25.0)/2.0, result.X, 1.0E-8);
````
may look better

**Comment by apdevelop** [comment]
Thank you for completed PR!

**Pull request merged**